tensorboard/webapp/widgets/data_table/filter_dialog_test.ts

```diff
@@ -0,0 +1,143 @@
+    const fixture = TestBed.createComponent(FilterDialog);
+    fixture.componentInstance.filter = input.filter;
+    rootLoader = TestbedHarnessEnvironment.documentRootLoader(fixture);
+    fixture.detectChanges();
+    return fixture;
+  }
+
+  it('dispatches event when an interval filter value is changed', async () => {
+    const fixture = createComponent({
+      filter: {
+        type: DomainType.INTERVAL,
+        includeUndefined: false,
+        minValue: 6,
+        maxValue: 20,
+        filterLowerValue: 7,
+        filterUpperValue: 18,
+      },
+    });
+
+    const intervalFilterChangedSpy = spyOn(
+      fixture.componentInstance.intervalFilterChanged,
+      'emit'
+    );
+
+    const rangeInput = fixture.debugElement.query(
+      By.directive(RangeInputComponent)
+    );
+    rangeInput.componentInstance.rangeValuesChanged.emit({
+      lowerValue: 7,
+      upperValue: 17,
+    });
+    expect(intervalFilterChangedSpy).toHaveBeenCalled();
+
+    const includeUndefinedSpy = spyOn(
+      fixture.componentInstance.includeUndefinedToggled,
+      'emit'
+    );
+
+    const includeUndefinedCheckbox = await rootLoader.getHarness(
+      MatCheckboxHarness.with({label: 'Include Undefined'})
+    );
+    await includeUndefinedCheckbox.check();
+    expect(includeUndefinedSpy).toHaveBeenCalled();
+  });
+
+  it('dispatches event when an discrete filter value is changed', async () => {
+    const fixture = createComponent({
+      filter: {
+        type: DomainType.DISCRETE,
+        includeUndefined: false,
+        possibleValues: [2, 4, 6, 8],
+        filterValues: [2, 4, 6],
+      },
+    });
+    const discreteFilterChangedSpy = spyOn(
+      fixture.componentInstance.discreteFilterChanged,
+      'emit'
+    );
+    const checkbox = await rootLoader.getHarness(
+      MatCheckboxHarness.with({label: '2'})
+    );
+    await checkbox.uncheck();
+    fixture.debugElement
+      .query(By.css('mat-checkbox label'))
+      .nativeElement.click();
+    expect(discreteFilterChangedSpy).toHaveBeenCalled();
+
+    const includeUndefinedSpy = spyOn(
+      fixture.componentInstance.includeUndefinedToggled,
+      'emit'
+    );
+
+    const includeUndefinedCheckbox = await rootLoader.getHarness(
+      MatCheckboxHarness.with({label: 'Include Undefined'})
+    );
+    await includeUndefinedCheckbox.check();
+    expect(includeUndefinedSpy).toHaveBeenCalled();
+  });
+
+  it('filters discrete values', async () => {
+    const fixture = createComponent({
+      filter: {
+        type: DomainType.DISCRETE,
+        includeUndefined: false,
+        possibleValues: ['foo', 'bar', 'baz', 'qaz'],
+        filterValues: ['foo', 'bar', 'baz', 'qaz'],
+      },
+    });
+    expect(
+      (await rootLoader.getAllHarnesses(MatCheckboxHarness)).length
+    ).toEqual(5); // 4 options + the include undefined checkbox
+
+    fixture.componentInstance.discreteValueFilter = 'ba';
```

Comment:
It seems like you are missing coverage of discreteValueKeyUp(). This seems like a good place to add that - instead of setting the discreteValueFilter variable directly can you instead type 'ba' into the box or somehow trigger the `change` event? If not, can you at least call discreteValueKeyUp() directly?

(Does using test harnesses make this easier to do?)